linalg/densemat.cpp

```diff
@@ -3500,4 +3502,114 @@ DenseTensor &DenseTensor::operator=(double c)
    return *this;
 }
 
+void BatchLUFactor(DenseTensor &Minv, Array<int> &P)
+{
+   const int m = Minv.SizeI();
+   const int NE = Minv.SizeK();
+   P.SetSize(m*NE);
+
+   auto data_all = mfem::Reshape(Minv.ReadWrite(), m, m, NE);
+   auto ipiv_all = mfem::Reshape(P.Write(), m, NE);
+   Array<bool> pivot_flag(1);
+   pivot_flag[0] = true;
+   bool *d_pivot_flag = pivot_flag.ReadWrite();
+   const double TOL = 1.e-9;
+
+   MFEM_FORALL(e, NE,
+   {
+      for (int i = 0; i < m; i++)
+      {
+         // pivoting
+         {
+            int piv = i;
+            double a = fabs(data_all(piv,i,e));
+            for (int j = i+1; j < m; j++)
+            {
+               const double b = fabs(data_all(j,i,e));
+               if (b > a)
+               {
+                  a = b;
+                  piv = j;
+               }
+            }
+            ipiv_all(i,e) = piv;
+            if (piv != i)
+            {
+               // swap rows i and piv in both L and U parts
+               for (int j = 0; j < m; j++)
+               {
+                  mfem::kernels::internal::Swap<double>(data_all(i,j,e), data_all(piv,j,e));
+               }
+            }
+         }//pivot end
+
+         if (abs(data_all(i,i,e)) <= TOL)
+         {
+            d_pivot_flag[0] = false;
+         }
+
+         const double a_ii_inv = 1.0 / data_all(i,i,e);
+         for (int j = i+1; j < m; j++)
+         {
+            data_all(j,i,e) *= a_ii_inv;
+         }
+
+         for (int k = i+1; k < m; k++)
+         {
+            const double a_ik = data_all(i,k,e);
+            for (int j = i+1; j < m; j++)
+            {
+               data_all(j,k,e) -= a_ik * data_all(j,i,e);
+            }
+         }
+
+      }//m loop
+
+   });
+
+   MFEM_ASSERT(pivot_flag.HostRead()[0], "Batch LU factorization failed \n");
+}
+
+void BatchLUSolve(const DenseTensor &Minv, const Array<int> &P, Vector &X)
+{
+
+   const int m = Minv.SizeI();
+   const int NE = Minv.SizeK();
+
+   auto data_all = mfem::Reshape(Minv.Read(), m, m, NE);
+   auto piv_all = mfem::Reshape(P.Read(), m, NE);
+   auto x_all = mfem::Reshape(X.ReadWrite(), m, NE);
+
+   MFEM_FORALL(e, NE,
+   {
+
+      // X <- P X
+      for (int i = 0; i < m; i++)
+      {
+         mfem::kernels::internal::Swap<double>(x_all(i,e), x_all(piv_all(i,e),e));
+      }
+
+      // X <- L^{-1} X
+      for (int j = 0; j < m; j++)
+      {
+         const double x_j = x_all(j,e);
+         for (int i = j+1; i < m; i++)
+         {
+            x_all(i,e) -= data_all(i,j,e) * x_j;
+         }
+      }
+
+      // X <- U^{-1} X
+      for (int j = m-1; j >= 0; j--)
+      {
+         const double x_j = ( x_all(j,e) /= data_all(j,j,e) );
+         for (int i = 0; i < j; i++)
+         {
+            x_all(i,e) -= data_all(i,j,e) * x_j;
+         }
+      }
```

Comment:
Replace this by a call to `LUSolve`, move `LUSolve` in this file, and modify `LUSolve` to use `DeviceTensor`?

Reply:
What about doing something this: 
```
   MFEM_FORALL(e, NE,                                                                                                                                                              
   {                                                                                                                                                                               
     kernels::LUSolve(&data_all(0, 0,e), &piv_all(0, e), &x_all(0,e));                                                                                                             
   });
```

The functions in ::kernels just use pointers and this would deviate for the format.